### ISSUE TITLE
Update Compiler Fork

### ISSUE THREAD

**Issue by thysultan** [zeit/styled-jsx#8]
There's a few things missing pre-v0.6.4 of the stylis compiler that your fork will not handle that could be nice to have.

1. flat css
```
/* this should get wrapped in the namespace */
color: red;

h1 {
    font-size:1px;
}
```

2. full prefix support in `@root and @keyframes blocks`. 

3. support opting out of name-spacing keyframes and animations.

I've done away with the regex solution i was using, and added supports for `stylis("[data-id=namespace]")` namespaces.

**Comment by rauchg** [comment]
@thysultan what do you think about `:host` support instead?

**Comment by thysultan** [comment]
@rauchg  So these will be the same`& { ... } === :host { ... }` ?

**Comment by rauchg** [comment]
yep so

```
<div>
  <style jsx>{`
    :host {
      color: red
    }
  `}</style>
</div>
```

`:host` would impact the div.

Alternatively, we could make the `div` also targetable simply with `div { }`, but I'm not sure if that's elegant

**Comment by rauchg** [comment]
cc @giuseppeg

**Comment by thysultan** [comment]
sure i can add that, `:host` will be an alias for `&`.

**Comment by giuseppeg** [comment]
@rauchg correct `:host` should resolve to the *shadow root* aka the `div` in the example. 

One thing to keep in mind is that styles applied from the outside should win over `:host`'s but I guess that this behavior is hard to shim.

Nice to have then it would be to get `:host-context(selector)` support e.g.:

```jsx
<body>
  <div>
    <style jsx>{`
      :host-context(body) {
        color: red
      }
    `}</style>
  </div>
</body>
```

yields

```css
body div[data-jsx="abc"] { color: red }
```

But probably it is better to tackle this later and maybe also consider [the `._aB1bC-ext` thing](https://github.com/zeit/next.js/issues/249#issuecomment-262178503) that I suggested in the original issue.

**Comment by thysultan** [comment]
@rauchg @giuseppeg pushed a patch for all the host variations `:host, :host(body), :host-context(body)`[commit](https://github.com/thysultan/stylis.js/commit/6efa027998daa107f14b0218b7c5b4440693dbe7#diff-5d9c96c5d55156f1be8b4a4d6a934a58R195)

**Comment by giuseppeg** [comment]
@thysultan cool in order for `* {}` to work you need to append the namespace though otherwise `${namespace}* {}` won't work.

**Comment by thysultan** [comment]
@giuseppeg  `* {}` will generate `${namespace} * {}` with a space.

**Comment by giuseppeg** [comment]
@thysultan oh I see, sorry I was thinking in terms of styled-jsx :)

**Comment by thysultan** [comment]
@rauchg instead of this

```
  <style jsx>{`
    :host {
      color: red
    }
  `}</style>
```

can this work

```
<style jsx>`
    :host {
      color: red
    }
`</style>
```

or better yet a plugin that treats everything in the <style jsx><style> block as a text node, removing the need for `\`\`` or `{}`

```
<style jsx>
    :host {
      color: red
    }
</style>
```

**Comment by rauchg** [comment]
The latter is invalid JSX unfortunately. It'll think of `{ … }` as an expression

**Comment by rauchg** [comment]
We picked the most optimal syntax considering how JSX works today I think :D

**Comment by rauchg** [comment]
@thysultan what happens here

https://github.com/zeit/styled-jsx/blob/master/lib/style-transform.js#L197

if the selector is something like:

```
[title="a,b"] {
  color: red
}
```

**Comment by thysultan** [comment]
@rauchg it will break, i fixed this in 0.7.1 [commit](https://github.com/thysultan/stylis.js/commit/d62d402ff261531937c9e9683b5fe6613377dd5e)

**Comment by giuseppeg** [comment]
@thysultan do you think that it would be trivial to switch to your version of the compiler and write a middleware to add prefixes the styled-jsx way? If so would you be interested in helping out?

**Comment by thysultan** [comment]
Sure, the middleware will look something like this

```js
function middleware (ctx, str, line, col, prefix) {
	if (ctx === 1.5) {
		if (str.indexOf(prefix) === 0) {
			return str.replace(prefix, '') + prefix;
		}
	}
}
```

This will change `h1, h2 {` to `h1[prefix], h2[prefix] {`

The `1.5` context is to be executed at every selector(post-processed) in a selector declaration.

**Comment by thysultan** [comment]
@giuseppeg On the repl site [https://stylis.js.org/](https://stylis.js.org/) You can try the following in the console.

```js
stylis.use(middleware (ctx, str, line, col, prefix) {
	if (ctx === 1.5) {
		if (str.indexOf(prefix+' ') === 0) {
			var pos = (str = str.substring(prefix.length).replace('&', '').trim()).indexOf(' ');
			return pos !== -1 ? str.substring(0, pos) + prefix + str.substring(pos) : str + prefix;
		}
	}
});
```

and input something into the editor and it should namespace the styled-jsx way.

**Comment by giuseppeg** [comment]
@thysultan the selector below should be `.logo[data-jsx="foo"] a[data-jsx="foo"]`

![screen shot 2017-02-10 at 5 06 56 pm](https://cloud.githubusercontent.com/assets/711311/22833745/5b35198e-efb3-11e6-9490-c73811465b50.png)

i.e. the prefix should be added to every part of the selector (except for globals). Do you think that it is doable with the current version of stylis + middleware?

**Comment by thysultan** [comment]
@giuseppeg Yes doable, what should `.logo:hover a {` compile to?

**Comment by giuseppeg** [comment]
cool `.logo[data-jsx="foo"]:hover a[data-jsx="foo"]`

**Comment by thysultan** [comment]
Then this should work as expected.

```js
function middleware (ctx, str, line, col, prefix) {
	if (ctx === 1.5) {
		if (str.indexOf(prefix+' ') === 0) {
			str = str.substring(prefix.length).trim();

			var pos = str.indexOf(' ');

			if (pos !== -1) {
				var parts = str.split(' ');

				for (var i = 0; i < parts.length; i++) {
					var part = parts[i];
					pos = part.indexOf(':');

					parts[i] = pos !== -1 ? part.substring(0, pos) + prefix + part.substring(pos) : part + prefix;
				}

				return parts.join(' ');
			}
			else {
				return str + prefix;
			}
		}
	}
}
```

**Comment by giuseppeg** [comment]
@thysultan @rauchg I started to work on this tonight https://github.com/zeit/styled-jsx/pull/125 keep in mind that it is very wip

**Comment by giuseppeg** [comment]
fixed by #134
